File: src/ObjectManager/Object.Tes/Formats/NifUtils.cs
﻿using OA.Core;
using UnityEngine;

namespace OA.Tes.Formats
{
    public static class NifUtils
    {
        public static Vector3 NifVectorToUnityVector(Vector3 nifVector)
        {
            Utils.Swap(ref nifVector.y, ref nifVector.z);
            return nifVector;
        }

        public static Vector3 NifPointToUnityPoint(Vector3 vifPoint)
        {
            return NifVectorToUnityVector(vifPoint) / ConvertUtils.MeterInUnits;
        }

        public static Matrix4x4 NifRotationMatrixToUnityRotationMatrix(Matrix4x4 nifRotationMatrix)
        {
            var matrix = new Matrix4x4
            {
                m00 = nifRotationMatrix.m00,
                m01 = nifRotationMatrix.m02,
                m02 = nifRotationMatrix.m01,
                m03 = 0,
                m10 = nifRotationMatrix.m20,
                m11 = nifRotationMatrix.m22,
                m12 = nifRotationMatrix.m21,
                m13 = 0,
                m20 = nifRotationMatrix.m10,
                m21 = nifRotationMatrix.m12,
                m22 = nifRotationMatrix.m11,
                m23 = 0,
                m30 = 0,
                m31 = 0,
                m32 = 0,
                m33 = 1
            };
            return matrix;
        }

        public static Quaternion NifRotationMatrixToUnityQuaternion(Matrix4x4 nifRotationMatrix)
        {
            return ConvertUtils.RotationMatrixToQuaternion(NifRotationMatrixToUnityRotationMatrix(nifRotationMatrix));
        }

        public static Quaternion NifEulerAnglesToUnityQuaternion(Vector3 nifEulerAngles)
        {
            var eulerAngles = NifVectorToUnityVector(nifEulerAngles);
            var xRot = Quaternion.AngleAxis(Mathf.Rad2Deg * eulerAngles.x, Vector3.right);
            var yRot = Quaternion.AngleAxis(Mathf.Rad2Deg * eulerAngles.y, Vector3.up);
            var zRot = Quaternion.AngleAxis(Mathf.Rad2Deg * eulerAngles.z, Vector3.forward);
            return xRot * zRot * yRot;
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/Formats/NifUtils.swift
//
//  NifReader.swift
//  ObjectManager
//
//  Created by Sky Morey on 6/5/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

import simd

public class NifUtils {
    
    public static func nifVectorToUnityVector(_ vector: Float3) -> Float3 {
        var unity = vector
        let z = unity.z
        unity.z = unity.y
        unity.y = z
        return unity
    }
    public static func nifVectorToUnityVector(_ vector: float3) -> float3 {
        var unity = vector
        let z = unity.z
        unity.z = unity.y
        unity.y = z
        return unity
    }
    
    public static func nifPointToUnityPoint(_ point: Float3) -> Float3 {
        return nifVectorToUnityVector(point) / ConvertUtils.meterInUnits
    }
    public static func nifPointToUnityPoint(_ point: float3) -> float3 {
        return nifVectorToUnityVector(point) / ConvertUtils.meterInUnits
    }

    public static func nifRotationMatrixToUnityRotationMatrix(_ rotationMatrix: simd_float4x4) -> simd_float4x4 {
        return simd_float4x4(
            float4(rotationMatrix[0].x, rotationMatrix[0].z, rotationMatrix[0].y, 0),
            float4(rotationMatrix[2].x, rotationMatrix[2].z, rotationMatrix[2].y, 0),
            float4(rotationMatrix[1].x, rotationMatrix[1].z, rotationMatrix[1].y, 0),
            float4(0, 0, 0, 1))
    }

    public static func nifRotationMatrixToUnityQuaternion(_ rotationMatrix: simd_float4x4) -> simd_quatf {
        return simd_quatf(nifRotationMatrixToUnityRotationMatrix(rotationMatrix))
    }

    public static func nifEulerAnglesToUnityQuaternion(_ nifEulerAngles: float3) -> simd_quatf {
        let eulerAngles2 = nifVectorToUnityVector(nifEulerAngles)
        let xRot = simd_quatf(angle: rad2Deg * eulerAngles2.x, axis: float3.right)
        let yRot = simd_quatf(angle: rad2Deg * eulerAngles2.y, axis: float3.up)
        let zRot = simd_quatf(angle: rad2Deg * eulerAngles2.z, axis: float3.forward)
        return xRot * zRot * yRot
    }
}
